File: Languages/cpp/algorithms-dsa/leetcode/dynamic-programming/longest_common_subsequence.cpp
/*
 * LeetCode 1143 - Longest Common Subsequence
 * Topic: Dynamic Programming
 * Difficulty: Medium
 *
 * Classic 2D DP: dp[i][j] = LCS of text1[0..i) and text2[0..j).
 * Time: O(m * n)
 * Space: O(m * n)
 */
#include <iostream>
#include <string>
#include <vector>
#include <cassert>
using namespace std;

int longestCommonSubsequence(string text1, string text2) {
    int m = text1.size(), n = text2.size();
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
    for (int i = 1; i <= m; i++)
        for (int j = 1; j <= n; j++)
            dp[i][j] = (text1[i-1] == text2[j-1])
                ? dp[i-1][j-1] + 1
                : max(dp[i-1][j], dp[i][j-1]);
    return dp[m][n];
}

int main() {
    assert(longestCommonSubsequence("abcde", "ace") == 3);
    assert(longestCommonSubsequence("abc", "abc") == 3);
    assert(longestCommonSubsequence("abc", "def") == 0);
    assert(longestCommonSubsequence("oxcpqrsvwf", "shmtulqrypy") == 2);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/go/algorithms-dsa/leetcode/dynamic-programming/longest_common_subsequence.go
/*
LeetCode #1143: Longest Common Subsequence
Topic: Dynamic Programming
Difficulty: Medium

Given two strings text1 and text2, return the length of their longest
common subsequence. A subsequence is derived by deleting characters
without changing relative order.

Approach: 2D DP table. If chars match, dp[i][j] = dp[i-1][j-1] + 1.
Otherwise dp[i][j] = max(dp[i-1][j], dp[i][j-1]).

Time: O(m * n), Space: O(m * n)
*/
package main

import "fmt"

func longestCommonSubsequence(text1 string, text2 string) int {
	m, n := len(text1), len(text2)
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if text1[i-1] == text2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = dp[i-1][j]
				if dp[i][j-1] > dp[i][j] {
					dp[i][j] = dp[i][j-1]
				}
			}
		}
	}
	return dp[m][n]
}

func main() {
	if longestCommonSubsequence("abcde", "ace") != 3 {
		fmt.Println("FAIL Test 1")
	} else {
		fmt.Println("PASS Test 1")
	}

	if longestCommonSubsequence("abc", "abc") != 3 {
		fmt.Println("FAIL Test 2")
	} else {
		fmt.Println("PASS Test 2")
	}

	if longestCommonSubsequence("abc", "def") != 0 {
		fmt.Println("FAIL Test 3")
	} else {
		fmt.Println("PASS Test 3")
	}
}
